### PR TITLE
HPCC-15565 Fix problems with inconsistent expression folding

### PR DESCRIPTION
**Pull request by ghalliday**
If something is folded when binding a parameter it should be folded
when a the ECL is parsed, otherwise it can lead to inconsistencies.

Signed-off-by: Gavin Halliday gavin.halliday@lexisnexis.com

### CODE REVIEW

**Comment by ghalliday** [comment]
@shamser please review

File: ecl/hql/hqlfold.cpp [hunk 1]
```diff
@@ -6354,8 +6354,8 @@ IHqlExpression * CExprFolderTransformer::transformExpanded(IHqlExpression * expr
 
 IHqlExpression * foldHqlExpression(IHqlExpression * expr)
 {
-    Owned<IErrorReceiver> errorProcessor = createNullErrorReceiver();
-    return foldHqlExpression(*errorProcessor, expr);
+    NullErrorReceiver errorProcessor;
+    return foldHqlExpression(errorProcessor, expr);
```
Comment (shamser) [inline ecl/hql/hqlfold.cpp]: Is this change necessary for this fix?

File: ecl/hql/hqlutil.cpp [hunk 2]
```diff
@@ -1293,8 +1293,8 @@ IHqlExpression * createImpureOwn(IHqlExpression * expr)
 
 IHqlExpression * getNormalizedFilename(IHqlExpression * filename)
 {
-    Owned<IErrorReceiver> errorProcessor = createNullErrorReceiver();
-    OwnedHqlExpr folded = foldHqlExpression(*errorProcessor, filename, NULL, HFOloseannotations);
+    NullErrorReceiver errorProcessor;
+    OwnedHqlExpr folded = foldHqlExpression(errorProcessor, filename, NULL, HFOloseannotations);
```
Comment (shamser) [inline ecl/hql/hqlutil.cpp]: Is this change necessary for this fix?

**Comment by shamser** [comment]
@ghalliday Just need clarification on couple of changes.  Everything else looks good.

File: ecl/hql/hqlfold.cpp (hunk 1 above)
Comment (ghalliday) [inline ecl/hql/hqlfold.cpp]: No this isn't necessary.  It is a minor optimization.

**Comment by ghalliday** [comment]
@richardkchapman see comments.  Is it ok to include the minor optimization in with the bug fix?  I can split it out if you prefer.
